Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp

```diff
@@ -269,14 +315,61 @@ void BackgroundFetch::Record::didSendData(uint64_t size)
         m_fetch->didSendData(size);
 }
 
+// https://wicg.github.io/background-fetch/#extract-content-range-values
+static ParsedContentRange extractContentRangeValues(const ResourceResponse& response)
+{
+    return ParsedContentRange(response.httpHeaderField(HTTPHeaderName::ContentRange));
+}
+
+// https://wicg.github.io/background-fetch/#validate-a-partial-response
+static bool validatePartialResponse(size_t rangeStart, const ResourceResponse& partialResponse, const ResourceResponse& previousResponse)
+{
+    auto parsedContentRange = extractContentRangeValues(partialResponse);
+    if (!parsedContentRange.isValid())
+        return false;
+
+    if (static_cast<size_t>(parsedContentRange.firstBytePosition()) != rangeStart)
+        return false;
+
+    auto previousETag = previousResponse.httpHeaderField(HTTPHeaderName::ETag);
+    if (!previousETag.isEmpty() && previousETag != partialResponse.httpHeaderField(HTTPHeaderName::ETag))
+        return false;
+
+    auto previousLastMotified = previousResponse.httpHeaderField(HTTPHeaderName::LastModified);
+    if (!previousLastMotified.isEmpty() && previousLastMotified != partialResponse.httpHeaderField(HTTPHeaderName::LastModified))
+        return false;
+
+    if (previousResponse.httpStatusCode() == 206) {
+        auto parsedPreviousContentRange = extractContentRangeValues(previousResponse);
+        if (!parsedPreviousContentRange.isValid())
+            return false;
+        if (parsedPreviousContentRange.instanceLength() != ParsedContentRange::unknownLength && parsedPreviousContentRange.instanceLength() != parsedContentRange.instanceLength())
+            return false;
+    }
+    return true;
+}
+
 void BackgroundFetch::Record::didReceiveResponse(ResourceResponse&& response)
 {
+    bool shouldClearResponseBody = false;
+    if (response.httpStatusCode() == 206) {
+        if (!validatePartialResponse(m_responseDataSize, response, m_response)) {
+            didFinish(ResourceError { String { }, 0, response.url(), "Validation of partial response failed"_s, ResourceError::Type::AccessControl });
+            return;
+        }
+    } else if (m_responseDataSize) {
+        m_responseDataSize = 0;
+        shouldClearResponseBody = true;
+        return;
```

Comment:
Seems either it should not return here or should not need to set `shouldClearResponseBody`

If you want to invoke `m_fetch->storeResponse` here, note that `m_responseDataSize` is updated to 0 and `m_currentDownloadSize` could be wrong (as `storeResponse` tries to read `responseDataSize()`)

Reply:
Right, return is not needed.